Release dependent modules only after the worker has finished for scheduled modules

makortel opened the pull request:
#### PR description:

In https://github.com/cms-sw/cmssw/issues/39064#issuecomment-1227671651 we found that the `PuttableProductResolver::putProduct()` releasing the task(s) for dependent modules can lead to rare problems if the producing module produces two (or more) products that refer to each other. Specifically, a consuming module may get run when only one of the products has been put in place, that can lead to deference of a Ref to fail on "missing product".

All consuming modules call the `prefetchAsync_()`, which for `PuttableProductResolver` already creates a task to release the dependent modules (`m_waitingTasks`), and inserts that into the Worker's `WaitingTaskList` (that is released after the Worker finishes). I think we could just rely on that ~~, and therefore this PR removes the override of `putProduct()` from `PuttableProductResolver`~~. This assumption turned out to be wrong, because there is also another use case for `PuttableProductResolver`: `PuttableSourceBase` sources (and `TestProcessor`, where it is also used like for sources).

As a simple way forward to fulfill the requirements of both use cases, this PR makes `PuttableProductResolver` to directly use the `Worker`'s `WaitingTaskList` in prefetching, if a `Worker` has been associated to the `PuttableProductResolver`. If no `Worker` has been associated, the `PuttableProductResolver` assumes it is from a Source (or similar), and the product is available for any consuming module without any further waiting in the prefetching. See also https://github.com/cms-sw/cmssw/pull/39245#issuecomment-1253916692 for some further discussion.

Resolves #39064 .

#### PR validation:

The reproducer I mentioned in https://github.com/cms-sw/cmssw/issues/39064#issuecomment-1228755900 succeeds with this change. Framework unit tests pass.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39245/31880

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/ProductResolvers.cc  modified in PR(s): #38454
  - File  FWCore/Framework/src/ProductResolvers.h  modified in PR(s): #38454

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

- FWCore/Framework (**core**)


@cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please review it and eventually sign? Thanks.
@wddgit this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
enable threading

makortel commented:
@cmsbuild, please test

makortel commented:
@Dr15Jones Would you agree?

Dr15Jones commented:
I'm not sure that works for the mixing module which doesn't (I think) do any prefetching.

Dr15Jones commented on FWCore/Framework/src/ProductResolvers.cc:
```diff
@@ -430,14 +430,6 @@ namespace edm {
-    ProducedProductResolver::putProduct(std::move(edp));
-    bool expected = false;
-    if (prefetchRequested_.compare_exchange_strong(expected, true)) {
-      m_waitingTasks.doneWaiting(std::exception_ptr());
```
The other `m_waitingTasks.doneWaiting` call requires that `worker_` not be `nullptr`. Is there any condition under which we do have it as `nullptr`?

makortel replied:
The `worker_` is initially uninitialized (already which sort of defeats the purpose of the check...) and is set later by
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/FWCore/Framework/src/ProductResolvers.cc#L449-L451

The `setupUnscheduled()` is called by `Principal::setupUnscheduled()`
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/FWCore/Framework/src/Principal.cc#L398-L400
that is called by
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/FWCore/Framework/src/WorkerManager.cc#L141-L148

That in turn is called from `GlobalSchedule::processOneGlobalAsync()`
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/FWCore/Framework/interface/GlobalSchedule.h#L220-L222
, from `StreamSchedule::processOneEventAsync()`
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/FWCore/Framework/src/StreamSchedule.cc#L915
and from
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/Mixing/Base/src/SecondaryEventProvider.cc#L121-L127
(the `SecondaryEventProvider::setupPileUpEvent()` ends up being called from `PreMixingModule` and `DataMixingModule` once for each pileup event)

To me it seems the workers are being set for the ProductResolvers before anything gets run (although I wonder why the setup gets repeated), and I can't see why anything would call `putProduct()` without `worker_` being set (but I could be missing some corner case).

makortel commented:
Do you mean the `SecondaryEventProvider` that is being used by `PileUp`
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/Mixing/Base/src/PileUp.cc#L106-L109 that is being used by `BMixingModule`
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/Mixing/Base/interface/BMixingModule.h#L108-L109
?

On a first look `SecondaryEventProvider` sets up the workers as unscheduled
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/Mixing/Base/src/SecondaryEventProvider.cc#L51-L67

But maybe there is some hidden detail that deserves deeper look. Unfortunately this won't be caught by any tests as the functionality has been unused for some years (and `Mixing/Base` has no tests).

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals RelVals-THREADING 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb5f3c/27179/summary.html
**COMMIT**: edbadbac2a40f3409333a55dd97ab67f8dc6354a
**CMSSW**: CMSSW_12_6_X_2022-08-29-1100/el8_amd64_gcc10
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39245/27179/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test TestFWCoreFrameworkTrigBit had ERRORS
</pre>

## RelVals

The relvals timed out after 4 hours.


## RelVals-THREADING

The relvals timed out after 4 hours.

makortel commented:
In addition, specifically HARVESTING step seems to get stuck (because of starvation of tasks in TBB)

Dr15Jones commented:
@makortel maybe the code that transitions from one module on a path to the next module on a path is adding to the module proxy's WaitingTaskList and since it is not a 'consumes' relation your change will fail? That it, it requires the release of the WaitingTaskList on the `put` if no module 'consumes' that data and it only runs because it is on a Path?

makortel committed:
Initialize stream workers before processing empty (End)Paths

makortel committed:
Use Worker's WaitingTaskList directly in PuttableProductResolver

This change was triggered by a case where a PuttableProductResolver
was filled by a Worker that produced many products, and one of the
products (A) had a Ref to another product (B), and the product B was
not consumed by any module (it was only accessed through the Ref).
Since the putProduct() released the WaitingTaskList of the Resolver,
that lead to the consumer of A to run, and that consumer dereferenced
the Ref to see that B was not there.

Besides scheduled modules another cases where PuttableProductResolver
is used are Sources inheriting PuttableSourceBase, and TestProcessor.
In these cases the products are put into the Resolvers (or left as
non-produced) before launching the prefetching of the unscheduled
system. Therefore in these use cases the consuming modules do not need
to wait for the Resolver to be filled.

After several fix attempts it seemed easiest to just use the Worker's
WaitingTaskList directly in PuttableProductResolver. This approach
fulfills the requirements of both Worker and Source(-like) use cases,
and even simplifies the code.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39245/32201

- This PR adds an extra 52KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/StreamSchedule.cc  modified in PR(s): #38983, #39409

cmsbuild commented:
Pull request #39245 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

makortel commented:
This push changes completely to the approach to one I discussed with @Dr15Jones yesterday. There are two use cases for `PuttableProductResolver`
- `Worker` corresponding a module in Path: the tasks for consuming modules should be released only after the worker has finished
- `PuttableSourceBase` and `TestProcessor`: products are put in advance (like by a module that is implicitly the first one in every Path), and therefore consuming modules do not need to wait

Using the `Worker`'s `WaitingTaskList` (if available) was the simplest solution to fulfill both behaviors, and allowed to also simplify the code. 

I tried also creating a separate ProductResolver tupe for the Source-like use case, and that would have worked as well, but would have required to add a way to communicate in the `BranchDescription` whether this new ProductResolver should be used. Making this approach nice would have required improving the current logic of identifying `PuttableSourceBase` sources (currently "if module label is `source`"), which would need quite some effort.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb5f3c/27716/summary.html
**COMMIT**: f88c583dd90b69cb2d1372a09c4c27c5f206c0d5
**CMSSW**: CMSSW_12_6_X_2022-09-21-1100/el8_amd64_gcc10
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39245/27716/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3624368
 - DQMHistoTests: Total failures: 11
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3624335
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-21-1100+cb5f3c/52918/triggerResults): no differences found

Dr15Jones commented on FWCore/Framework/src/ProductResolvers.h:
```diff
@@ -218,12 +218,12 @@ namespace edm {
-    CMS_THREAD_SAFE mutable WaitingTaskList m_waitingTasks;
-    Worker* worker_;
-    mutable std::atomic<bool> prefetchRequested_;
+    // The WaitingTaskList below is the one of the worker, if one
```
```suggestion
    // The WaitingTaskList below is the one from the worker, if one
```

makortel replied:
Thanks, done

Dr15Jones commented on FWCore/Framework/interface/maker/Worker.h:
```diff
@@ -241,6 +241,9 @@ namespace edm {
     virtual bool hasAccumulator() const = 0;
 
+    // Used in PuttableProductResolver
+    edm::WaitingTaskList* waitingTaskList() { return &waitingTasks_; }
```
Given this can never return a null, maybe it would be better to return `edm::WaitingTaskList&`

makortel replied:
Thanks, done

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39245/32217

- This PR adds an extra 40KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/StreamSchedule.cc  modified in PR(s): #38983, #39409

cmsbuild commented:
Pull request #39245 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb5f3c/27739/summary.html
**COMMIT**: e8a6e6ab328e3c963c3e5fa4e2508d2111e62f37
**CMSSW**: CMSSW_12_6_X_2022-09-22-1100/el8_amd64_gcc10
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39245/27739/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3624368
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3624341
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-22-1100+cb5f3c/52931/triggerResults): no differences found

makortel commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

makortel commented:
type bug-fix

perrotta commented:
+1
- Fully trusting the gurus who contributed to this PR
- PR tests pass successfully
